src/Raven.Server/Documents/Sharding/Background/ShardedPeriodicDocumentsMigrator.cs

```diff
@@ -0,0 +1,95 @@
+
+            if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                return;
+
+            await ExecuteMoveDocuments();
+        }
+
+        internal async Task ExecuteMoveDocuments()
+        {
+            try
+            {
+                int bucket = -1;
+                int moveToShard = -1;
+                bool found = false;
+                using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    var configuration = _database.ShardingConfiguration;
+                    for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                    {
+                        var range = configuration.BucketRanges[index];
+                        if (range.ShardNumber == _database.ShardNumber)
+                            continue;
+
+                        var start = range.BucketRangeStart;
+                        var end = index == configuration.BucketRanges.Count - 1
+                            ? int.MaxValue
+                            : configuration.BucketRanges[index + 1].BucketRangeStart - 1;
+
+                        var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                        foreach (var bucketStats in bucketStatistics)
+                        {
+                            if (bucketStats.NumberOfDocuments == 0)
+                                continue;
+
+                            bucket = bucketStats.Bucket;
+                            moveToShard = range.ShardNumber;
+                            found = true;
+                            break;
+                        }
+
+                        if (found)
+                            break;
+                    }
+                }
+
+                if (found)
+                    await MoveDocumentsToShard(bucket, moveToShard);
+            }
+            catch (OperationCanceledException)
+            {
+                // this will stop processing
+                throw;
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations($"Failed to execute documents migration for '{_database.Name}'", e);
+
+                throw;
+            }
+        }
+
+        private async Task MoveDocumentsToShard(int bucket, int moveToShard)
+        {
+            var cmd = new StartBucketMigrationCommand(bucket, _database.ShardNumber, moveToShard, _database.ShardedDatabaseName,
+                $"{Guid.NewGuid()}/{bucket}");
+
+            var result = await _database.ServerStore.SendToLeaderAsync(cmd);
+            await _database.ServerStore.Cluster.WaitForIndexNotification(result.Index);
```

Comment:
This doesn't wait for the actual operation to complete? 
Just until it is ack by the cluster.
If we have 10 buckets to move, it would do:

min 1: - start migration bucket 1, wait 60 sec
min 2: - migration is running, wait 60 sec
min 3: - start migration bucket 2, wait 60 sec

etc

That is going to be very slow, no?